### ISSUE TITLE
Problems with System.Text.Json on .NET8

### ISSUE THREAD

**Issue by inf9144** [6bee/aqua-core#41]
Hey,
we are using Remote.Linq and Aqua.DynamicObject for a long time and really like it.
It crashes on .NET 8 with the following message:

```
    System.NotSupportedException : The converter for derived type 'Aqua.Dynamic.DynamicObject' does not support metadata writes or reads. The unsupported member type is located on type 'System.Dynamic.IDynamicMetaObjectProvider'. Path: $.
    ---- System.NotSupportedException : The converter for derived type 'Aqua.Dynamic.DynamicObject' does not support metadata writes or reads.
```

Have these configuration flags set:
```
         MaxDepth = int.MaxValue,
         ReferenceHandler = SystemTextReferenceHandler.Preserve,
         NumberHandling = SystemTextJsonNumberHandling.AllowNamedFloatingPointLiterals | SystemTextJsonNumberHandling.AllowReadingFromString,
         WriteIndented = true,
         DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault,
```

Can you point me into the right direction? Maybe we can fix it and contribute as PR ;-)

Best Regards,
Alex

**Comment by inf9144** [comment]
Cant reproduce it when checking out aqua-core. Seems to have something to do with my JsonPolymorphismOptions in one of our TypeResolvers. But Aqua.Core has a problem on .NET 8 since FormatterServices are used (UtilizeFormatterServices is true in default settings) and those got deprecated. Will look further into this.

**Comment by 6bee** [comment]
Hi,
Thanks for reporting this. I'll look into it.

**Comment by inf9144** [comment]
Hey thanks for your fast response :-)
Having the next thing that was changed in System.Text.Json - reading of metadata. 
Right now im running in the case where TryReadMetadata of JsonSerializer is checking the $type token which is set on a non-polymorphic type (struct), which they disallowed here:

```
            if (jsonTypeInfo.PolymorphicTypeResolver is null)
                            {
                                // Found a $type property in a type that doesn't support polymorphism
                                ThrowHelper.ThrowJsonException_MetadataUnexpectedProperty(propertyName, ref state);
                            }
```

but the ObjectConverter<T> of aqua is writing it in all cases (which i really like ;-) )
Will check if it's possible to get around this - i would think that JsonSerializer shouldnt look into it as a converter is registered.

//Edit:
It's not the converter writing the $type - it's System.Text.Json.JsonSerializer itself - need to check how i got into that situation. :-/

**Comment by inf9144** [comment]
I think i found it.

On the writer site you are calling:
`writer.Serialize(property.Value, options);`
which makes the JsonSerializer serialize a `<T>` which is `<object>`
that causes support for polymorphism (I could use an interface as type and have many structs that implement it)
Thats why it writes $type

On the reader site you are calling:
`reader.Read(type, options);`
Here the struct type is already mapped from the "Type" property and the serializer now knows that there is no need for polymorphism support and ends up in throwing the error.

I think the writer side should do:
`writer.Serialize(property.Value, property.Value.GetType(), options);`
than both should be symmetric and it shouldnt cause JsonSerializer to emit $type

Would program a little repro - but it's to late right now. Maybe monday ;-)

//Edit: Only happens if the interface's TypeInfo gets PolymorphismOptions set where DerivedTypes contains the structs. If there is no type with PolymorphismOptions that says that the structs are it's derived types, the serializer call does not emit $type. The fix above should work :-) The check if polymorphy needed is skipped if `<T>` / type is provided and is ValueType.

**Comment by 6bee** [comment]
Hi Alex,
Thank you for all the information you provided. 
Meanwhile I've added a .NET 8.0 target to remove binary serialization for .NET 8.0 and later. 
Unfortunately, I was unable to reproduce the issues you reported. 
I've created a small unit test for trying to reproduce the issue (see commit d90db4e9208952dc12da6da460dd41aab5c5d407, branch `issues/41-repro`). 
It would be much appreciated if you could have a look to help reproduce it, either using _aqua-core_ or _remote-linq_.

**Comment by inf9144** [comment]
Hi Christof,
i will check out the branch and will try to create a reprodruction on the weekend :-)

**Comment by inf9144** [comment]
Reproduced it! :-)

Just copy this testcase (sorry couldnt bring my visual studio to comply to your code style and could not fit it in your test hierarchy because of all the modifications to the serializer settings):

```
#pragma warning disable SA1200
#pragma warning disable SA1500
#pragma warning disable SA1413

using Aqua.Dynamic;
using Aqua.Text.Json;
using Shouldly;
using System;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;
using Xunit;

namespace Aqua.Tests.Serialization.Dynamic.DynamicObject;

public class TestWithInterfaceOnStruct {
   public interface IInterfaceForStructs {
      public int Value { get; init; }
   }

   public readonly struct StructA : IInterfaceForStructs {
      public int Value { get; init; }

      public StructA() {
      }
   }

   public readonly struct StructB : IInterfaceForStructs {
      public int Value { get; init; }

      public StructB() {
      }
   }

   public class ObjectWithStruct {
      public IInterfaceForStructs SomeStruct { get; init; }

      public ObjectWithStruct() {
      }
   }

   private sealed class IsKnownTypeProvider : IIsKnownTypeProvider {
      public bool IsKnownType(Type type)
          => type == typeof(StructA) || type == typeof(StructB) || type == typeof(IIsKnownTypeProvider);
   }

   [Fact]
   public void TestSerializationWithPolymorphInterfaceOnStructs() {
      var serializerOptions = new JsonSerializerOptions() {
         WriteIndented = true,
         ReferenceHandler = ReferenceHandler.Preserve,
         TypeInfoResolver = new DefaultJsonTypeInfoResolver() {
            Modifiers = {
               AddPolymorphy
            }
         }
      }.ConfigureAqua();

      var originalObject = new ObjectWithStruct() { SomeStruct = new StructA() { Value = 12 } };
      var serialized = JsonSerializer.Serialize(new DynamicObjectMapper(isKnownTypeProvider: new IsKnownTypeProvider()).MapObject(originalObject), serializerOptions);
      var copy = new DynamicObjectMapper(isKnownTypeProvider: new IsKnownTypeProvider()).Map<ObjectWithStruct>(JsonSerializer.Deserialize<Aqua.Dynamic.DynamicObject>(serialized, serializerOptions));
      copy.SomeStruct.Value.ShouldBe(originalObject.SomeStruct.Value);
   }

   private static void AddPolymorphy(JsonTypeInfo typeInfo) {
      if (typeInfo.Type == typeof(IInterfaceForStructs)) {
         typeInfo.PolymorphismOptions ??= new();
         typeInfo.PolymorphismOptions.IgnoreUnrecognizedTypeDiscriminators = true;
         typeInfo.PolymorphismOptions.UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor;
         typeInfo.PolymorphismOptions.DerivedTypes.Add(new JsonDerivedType(typeof(StructA), typeof(StructA).FullName));
         typeInfo.PolymorphismOptions.DerivedTypes.Add(new JsonDerivedType(typeof(StructB), typeof(StructB).FullName));
      }
   }
}
```

If you change line 288 in DynamicObjectConverter from:
`writer.Serialize(property.Value, options);`
to
`writer.Serialize(property.Value, property.Value.GetType(), options);`
The test (and all other tests) get green :-)

Thanks for your patience - weekend was busier than expected ;-)

**Comment by 6bee** [comment]
Thanks so much for providing a repro!

As I understand, the issue is caused with _PolymorphismOptions_ defined and **not** passing the type argument  to _serialize_.

I.e. the _type discriminator_ metadata is not added to json when passing the type argument to _serialize_!? In other words, the issue is resolved as _PolymorphismOptions_ are actually not used/applied when the type is specified. Also, if no _PolymorphismOptions_ are declared in the first place, the issues doesn't happen either.

I don't fully understand the mechanism here. However, I'm happy to apply your fix as it doesn't seem to break anything else. 

The fix can be tested using `aqua-core 5.4.0-alpha-002` and/or `Remote.Linq 7.2.0-alpha-003`.

**Comment by 6bee** [comment]
Closing issue due to inactivity. Feel free to re-open if required.
